src/activities/Elsa.Activities.Timers/HostedServices/TimersHostedService.cs

```diff
@@ -29,6 +30,12 @@ public class TimersHostedService : BackgroundService
             this.logger = logger;
         }
 
+        public override async Task StartAsync(CancellationToken cancellationToken)
+        {
+            await distributedLockProvider.SetupAsync(cancellationToken);
```

Comment:
What would you say if we did the following:

1. Remove the `SetupAsync`.
1. Implement a `SetupRedisLockProvider: IHostedService` class that takes care of the setup autimatically.

Couple of advantages with that:

1. No need for each background task that needs distributred locking (like this `TimersHostedService`) to imperatively call `SetupAsync` - which might be inefficient if it is called multiple times, plus if someone implementing their own activity services forgets about this, it becomes inconsistent and prone to error.
1. No need anymore to "force" a `SetupAsync` on other distributed locking providers.

Conceptually, it would be akin to having a hosted servie that takes care of setting up database indices, running migrations etc - an async setup of the system before ASP.NET Core starts serving requests.

Thoughts?

Reply:
That makes sense! however i will have to look into it. 

If user wants to develop custom provider,how can ELSA enforces to implement that any LockProvider(in our cases) should also implement IHostedService? 

Also, If we do that connection will stay open a long as application is not stopped -- Happy Path 
In case of transient issue b/w ELSA and Redis, if connection broke, how it will up again? 

am i missing something?

Reply:
@sfmskywalker  I guess `SetupAsync `name is misleading, this method only connects to redis, you might need to take a look on that. 

[https://github.com/anktsrkr/elsa-core/blob/feature/elsa-2.0redisdistributedlockprovider/src/providers/Elsa.DistributedLocking.Redis/RedisLockProvider.cs#L73](https://github.com/anktsrkr/elsa-core/blob/feature/elsa-2.0redisdistributedlockprovider/src/providers/Elsa.DistributedLocking.Redis/RedisLockProvider.cs#L73)

My suggestion will be keep the method there but rename it. 

Thoughts?

Reply:
@anktsrkr When a user wants to develop a custom provider that needs some hosted service to do some setup, I imagine this provider will ship with an extension method (on IServiceCollection), like the other modules do. In this extension method, all the necessary services would be registered: the provider as well as the hosted service. This is what I do for example in Elsa.Timers, where not only new activities are registered, but also a hosted service.

Reply:
> Also, If we do that connection will stay open a long as application is not stopped -- Happy Path
In case of transient issue b/w ELSA and Redis, if connection broke, how it will up again?

This depends on how you set things up. For example, if all you need to do is setup something on the service one time, you can do this work in the Start method of the hosted service (just implement IHostedService directly instead of BackgroundService).

Re-establishing lost Redis connections could be done in the background (using a hosted service), or perhaps on the fly, when e.g. you try to read something from Redis. Depending on the library you're using, reconnecting should be handled for you according to [this post](https://stackoverflow.com/questions/24922782/recovering-from-redis-connectivity-loss).

Reply:
Indeed we do not need `SetupAsync`. Even we don't need `IHostedService` , StackExchange.Redis is capable enough to handle Transient error and ELSA will wait for 60 seconds to acquire it. In case of permanent server shutdown, ELSA will be unable to acquire lock and it will return false which is expected.  

I am still checking for possible enhancement, but hopefully by tommrow I can commit my changes.

Reply:
Sounds good. I checked your changes and it looks great. Thanks!